### PR TITLE
add BiMap / MultiBiMap / AsymmetricBiMap / Caches + rename Auxilary.Utils -> Auxilary.References

### PR DESCRIPTION
**Pull request by aria1th**

### CODE REVIEW

**Comment by aria1th** [comment]
Now we have circular dependency :panic:

**Comment by senseiwells** [comment]
Where?

**Comment by aria1th** [comment]
MultiBiMap tries to use VarArgs, which is from Auxilary
but Auxilary uses ImmutableList from Collections

I'll fix it to not use VarArgs, but will you fix those circular dependency?

**Comment by senseiwells** [comment]
There's no way to 'fix' circular dependancies

**Comment by senseiwells** [comment]
However you can just move you import to later in the file

**Comment by senseiwells** [comment]
For example:

File 'A.arucas'
```kotlin
import ClassB from B;

class ClassA {
}

new B();
```

File 'B.arucas'
```kotlin
classB {
}

import ClassA from A;

new A();
```

File: libs/util/Auxilary.arucas [hunk 1]
```diff
@@ -1,27 +1,34 @@
...
 
 /**
- * Miscilaneous methods that don't fit into other categories.
+ * Miscilaneous methods that dont fit into other categories.
  */
-class Util {
-    Util() {
-        throw Error();
+
+class References {
```
Comment (senseiwells) [inline libs/util/Auxilary.arucas:7]: `Reference` might be a better name?

File: libs/util/Auxilary.arucas [hunk 2]
```diff
@@ -1,27 +1,34 @@
-import AssertionError from util.Errors;
-import ImmutableList from util.Collections;
+import IllegalAccessError, AssertionError from util.Errors;
 
 /**
- * Miscilaneous methods that don't fit into other categories.
+ * Miscilaneous methods that dont fit into other categories.
```
Comment (senseiwells) [inline libs/util/Auxilary.arucas:4]: Why was this changed lol.
I mean don't is correct.

File: libs/util/Auxilary.arucas [hunk 3]
```diff
@@ -57,6 +70,11 @@ class Assert {
         throw new AssertionError("No error was thrown");
     }
 }
+import IndexError, TypeError from util.Errors;
+import ImmutableList from util.Collections;
```
Comment (senseiwells) [inline libs/util/Auxilary.arucas:74]: Imports can be moved to the top (with new lazy imports)

File: libs/util/Auxilary.arucas [hunk 4]
```diff
@@ -69,25 +87,37 @@ class VarArgs {
...
-        value = this.get(index);
-        if (value.instanceOf(type)) {
-            return value;
+    fun getOrDefault(index: Number, defaultValue: Object): Object {
+        value = defaultValue;
+        try {
+            value = this.get(index);
+        }
+        catch (error) {
+            if (error.instanceOf(IndexError.type)) return value;
```
Comment (senseiwells) [inline libs/util/Auxilary.arucas:109]: This is kinda weird. Why not just check the size, and if it's out of bounds then return default instead of doing this weird catch check.

File: libs/util/Caches.arucas [hunk 5]
```diff
@@ -0,0 +1,75 @@
+import Assert from util.Auxilary;
+import AsymmetricBiMap from util.Collections;
+/*
+*   Class Cache - basic Cache object that removes items sequentially
+*/
+class Cache : Map {
+    var map : Map;
+    var MAX_SIZE : Number = 16;
```
Comment (senseiwells) [inline libs/util/Caches.arucas:8]: WHY IS THIS CAPITAL?

File: libs/util/Caches.arucas [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+    fun put(key, value) {
+        if (!this.containsKey(key)) {
+            super.put(key, value);
+            this.checkMAX_SIZE();
+        }
+        else {
+            super.put(key, value);
+        }
+    }
+    fun checkMAX_SIZE() {
```
Comment (senseiwells) [inline libs/util/Caches.arucas:26]: FUNCTIONS SHOULD FOLLOW CAMEL CASING.

**Review by senseiwells — CHANGES_REQUESTED** [review]
Pretty good, just some minor changes

File: libs/util/Auxilary.arucas (hunk 2 above)
Comment (aria1th) [reply libs/util/Auxilary.arucas:4]: IntelliJ is suffering for that somehow, because /** format is not recognized as comments....

File: libs/util/Caches.arucas (hunk 6 above)
Comment (aria1th) [reply libs/util/Caches.arucas:26]: notepad moment oof
I forgot to change it to capacity

File: libs/util/Auxilary.arucas (hunk 4 above)
Comment (aria1th) [reply libs/util/Auxilary.arucas]: yup in this case no other error is thrown except index error, but I feel like there should be proper way to catch certain error types

File: libs/util/Collections.arucas [hunk 7]
```diff
@@ -692,8 +692,8 @@ class BiMap: Map {
         if (this.containsKey(key) && value != this.get(key)) {
             throw new IllegalArgumentError("Cannot input duplcate key " + key + " for value : "+ value + ", use BiMap.forcePut(key, value) to replace!");
         }
-        if (this.containsValue(value) && key != this.getKey(value)) {
-            throw new IllegalArgumentError("Cannot input duplcate value " + value + " for key : "+ key + ", use BiMap.forcePut(key, value) to replace!");    
+        if (this.containsValue(value) && key != this.getKeys(value)) {
```
Comment (senseiwells) [inline libs/util/Collections.arucas:695]: This should be `getKey(value)` gets the key for a value (just a method that wraps `this.back.get(value)`.
`getKeys` is incorrect

File: libs/util/Collections.arucas [hunk 8]
```diff
@@ -812,7 +812,7 @@ class AsymmetricBiMap : BiMap {
             this.front.put(key, value);
         }
         if (this.containsValue(value)) {
-            this.getKey(value).add(key);
+            this.getKeys(value).add(key);
```
Comment (senseiwells) [inline libs/util/Collections.arucas:815]: This should be `getKey(value)` gets the key for a value (just a method that wraps `this.back.get(value)`.
`getKeys` is incorrect

**Review by senseiwells — COMMENTED** [review]
Minor issue
